In [142]:
!pip install --upgrade pip
!pip install beautifulsoup4
!pip install requests
!pip install lxml
!pip install geopy
!pip install sklearn
!pip install scikit-learn
!pip install folium


Defaulting to user installation because normal site-packages is not writeable
Defaulting to user installation because normal site-packages is not writeable
Defaulting to user installation because normal site-packages is not writeable
Defaulting to user installation because normal site-packages is not writeable
Defaulting to user installation because normal site-packages is not writeable
Defaulting to user installation because normal site-packages is not writeable
Defaulting to user installation because normal site-packages is not writeable
Defaulting to user installation because normal site-packages is not writeable


# Data processing and Cleaning of the Postal Codes of Canada

In [323]:
import requests
import lxml.html as lh
import bs4 as bs
import urllib.request
import folium
import matplotlib.pyplot as plt # plotting library
import matplotlib.cm as cm
import matplotlib.colors as colors
# import k-means from clustering stage
from sklearn.cluster import KMeans

import numpy as np # library to handle data in a vectorized manner
import pandas as pd # library for data analsysis

pd.set_option('display.max_columns', None)
pd.set_option('display.max_rows', None)

import json # library to handle JSON files


from geopy.geocoders import Nominatim 
GeoLocator = Nominatim(user_agent='My-IBMNotebook')# convert an address into latitude and longitude values

In [324]:
import numpy as np # library to handle data in a vectorized manner
import pandas as pd # library for data analsysis

pd.set_option('display.max_columns', None)
pd.set_option('display.max_rows', None)

import json # library to handle JSON files


from geopy.geocoders import Nominatim 

GeoLocator = Nominatim(user_agent='My-Application')# convert an address into latitude and longitude values

import requests # library to handle requests
from pandas.io.json import json_normalize # tranform JSON file into a pandas dataframe

# Matplotlib and associated plotting modules
import matplotlib.cm as cm
import matplotlib.colors as colors

# import k-means from clustering stage
from sklearn.cluster import KMeans


import folium # map rendering library

print('Libraries imported.')

Libraries imported.


In [325]:
url = "https://en.wikipedia.org/wiki/List_of_postal_codes_of_Canada:_M"
res = requests.get(url)
soup = bs.BeautifulSoup(res.content,'lxml')

# Data Processing

In [326]:
table_contents=[]
table=soup.find('table')
for row in table.findAll('td'):
    cell = {}
    if row.span.text=='Not assigned':
        pass
    else:
        cell['PostalCode'] = row.p.text[:3]
        cell['Borough'] = (row.span.text).split('(')[0]
        cell['Neighborhood'] = (((((row.span.text).split('(')[1]).strip(')')).replace(' /',',')).replace(')',' ')).strip(' ')
        table_contents.append(cell)


# Data Cleaning and printing the data

In [329]:
#print(table_contents)
df=pd.DataFrame(table_contents)
df['Borough']=df['Borough'].replace({'Downtown TorontoStn A PO Boxes25 The Esplanade':'Downtown Toronto Stn A',
                                             'East TorontoBusiness reply mail Processing Centre969 Eastern':'East Toronto Business',
                                             'EtobicokeNorthwest':'Etobicoke Northwest','East YorkEast Toronto':'East York/East Toronto',
                                             'MississaugaCanada Post Gateway Processing Centre':'Mississauga'})
df

PostalCode                 Borough  \
0          M3A              North York   
1          M4A              North York   
2          M5A        Downtown Toronto   
3          M6A              North York   
4          M7A            Queen's Park   
5          M9A               Etobicoke   
6          M1B             Scarborough   
7          M3B              North York   
8          M4B               East York   
9          M5B        Downtown Toronto   
10         M6B              North York   
11         M9B               Etobicoke   
12         M1C             Scarborough   
13         M3C              North York   
14         M4C               East York   
15         M5C        Downtown Toronto   
16         M6C                    York   
17         M9C               Etobicoke   
18         M1E             Scarborough   
19         M4E            East Toronto   
20         M5E        Downtown Toronto   
21         M6E                    York   
22         M1G             Scarborough   
23         M4G               East York   
24         M5G        Downtown Toronto   
25         M6G        Downtown Toronto   
26         M1H             Scarborough   
27         M2H              North York   
28         M3H              North York   
29         M4H               East York   
30         M5H        Downtown Toronto   
31         M6H            West Toronto   
32         M1J             Scarborough   
33         M2J              North York   
34         M3J              North York   
35         M4J  East York/East Toronto   
36         M5J        Downtown Toronto   
37         M6J            West Toronto   
38         M1K             Scarborough   
39         M2K              North York   
40         M3K              North York   
41         M4K            East Toronto   
42         M5K        Downtown Toronto   
43         M6K            West Toronto   
44         M1L             Scarborough   
45         M2L              North York   
46         M3L              North York   
47         M4L            East Toronto   
48         M5L        Downtown Toronto   
49         M6L              North York   
50         M9L              North York   
51         M1M             Scarborough   
52         M2M              North York   
53         M3M              North York   
54         M4M            East Toronto   
55         M5M              North York   
56         M6M                    York   
57         M9M              North York   
58         M1N             Scarborough   
59         M2N              North York   
60         M3N              North York   
61         M4N         Central Toronto   
62         M5N         Central Toronto   
63         M6N                    York   
64         M9N                    York   
65         M1P             Scarborough   
66         M2P              North York   
67         M4P         Central Toronto   
68         M5P         Central Toronto   
69         M6P            West Toronto   
70         M9P               Etobicoke   
71         M1R             Scarborough   
72         M2R              North York   
73         M4R         Central Toronto   
74         M5R         Central Toronto   
75         M6R            West Toronto   
76         M7R             Mississauga   
77         M9R               Etobicoke   
78         M1S             Scarborough   
79         M4S         Central Toronto   
80         M5S        Downtown Toronto   
81         M6S            West Toronto   
82         M1T             Scarborough   
83         M4T         Central Toronto   
84         M5T        Downtown Toronto   
85         M1V             Scarborough   
86         M4V         Central Toronto   
87         M5V        Downtown Toronto   
88         M8V               Etobicoke   
89         M9V               Etobicoke   
90         M1W             Scarborough   
91         M4W        Downtown Toronto   
92         M5W  Downtown Toronto Stn A   
93         M8W               Etobicoke   
94         M9W

# shape of the data

In [330]:
df.shape

(103, 3)

# Geo coordinates of the data
# Reading the csv file containing the coordinates

In [265]:
df_geo=pd.read_csv('Geospatial_Coordinates.csv')
df_geo

Postal Code   Latitude  Longitude
0           M1B  43.806686 -79.194353
1           M1C  43.784535 -79.160497
2           M1E  43.763573 -79.188711
3           M1G  43.770992 -79.216917
4           M1H  43.773136 -79.239476
5           M1J  43.744734 -79.239476
6           M1K  43.727929 -79.262029
7           M1L  43.711112 -79.284577
8           M1M  43.716316 -79.239476
9           M1N  43.692657 -79.264848
10          M1P  43.757410 -79.273304
11          M1R  43.750072 -79.295849
12          M1S  43.794200 -79.262029
13          M1T  43.781638 -79.304302
14          M1V  43.815252 -79.284577
15          M1W  43.799525 -79.318389
16          M1X  43.836125 -79.205636
17          M2H  43.803762 -79.363452
18          M2J  43.778517 -79.346556
19          M2K  43.786947 -79.385975
20          M2L  43.757490 -79.374714
21          M2M  43.789053 -79.408493
22          M2N  43.770120 -79.408493
23          M2P  43.752758 -79.400049
24          M2R  43.782736 -79.442259
25          M3A  43.753259 -79.329656
26          M3B  43.745906 -79.352188
27          M3C  43.725900 -79.340923
28          M3H  43.754328 -79.442259
29          M3J  43.767980 -79.487262
30          M3K  43.737473 -79.464763
31          M3L  43.739015 -79.506944
32          M3M  43.728496 -79.495697
33          M3N  43.761631 -79.520999
34          M4A  43.725882 -79.315572
35          M4B  43.706397 -79.309937
36          M4C  43.695344 -79.318389
37          M4E  43.676357 -79.293031
38          M4G  43.709060 -79.363452
39          M4H  43.705369 -79.349372
40          M4J  43.685347 -79.338106
41          M4K  43.679557 -79.352188
42          M4L  43.668999 -79.315572
43          M4M  43.659526 -79.340923
44          M4N  43.728020 -79.388790
45          M4P  43.712751 -79.390197
46          M4R  43.715383 -79.405678
47          M4S  43.704324 -79.388790
48          M4T  43.689574 -79.383160
49          M4V  43.686412 -79.400049
50          M4W  43.679563 -79.377529
51          M4X  43.667967 -79.367675
52          M4Y  43.665860 -79.383160
53          M5A  43.654260 -79.360636
54          M5B  43.657162 -79.378937
55          M5C  43.651494 -79.375418
56          M5E  43.644771 -79.373306
57          M5G  43.657952 -79.387383
58          M5H  43.650571 -79.384568
59          M5J  43.640816 -79.381752
60          M5K  43.647177 -79.381576
61          M5L  43.648198 -79.379817
62          M5M  43.733283 -79.419750
63          M5N  43.711695 -79.416936
64          M5P  43.696948 -79.411307
65          M5R  43.672710 -79.405678
66          M5S  43.662696 -79.400049
67          M5T  43.653206 -79.400049
68          M5V  43.628947 -79.394420
69          M5W  43.646435 -79.374846
70          M5X  43.648429 -79.382280
71          M6A  43.718518 -79.464763
72          M6B  43.709577 -79.445073
73          M6C  43.693781 -79.428191
74          M6E  43.689026 -79.453512
75          M6G  43.669542 -79.422564
76          M6H  43.669005 -79.442259
77          M6J  43.647927 -79.419750
78          M6K  43.636847 -79.428191
79          M6L  43.713756 -79.490074
80          M6M  43.691116 -79.476013
81          M6N  43.673185 -79.487262
82          M6P  43.661608 -79.464763
83          M6R  43.648960 -79.456325
84          M6S  43.651571 -79.484450
85          M7A  43.662301 -79.389494
86          M7R  43.636966 -79.615819
87          M7Y  43.662744 -79.321558
88          M8V  43.605647 -79.501321
89          M8W  43.602414 -79.543484
90          M8X  43.653654 -79.506944
91          M8Y  43.636258 -79.498509
92          M8Z  43.628841 -79.520999
93          M9A  43.667856 -79.532242
94          M9B  43.650943 -79.554724
95          M9C  43.643515 -79.577201
96          M9L  43.756303 -79.565963
97          M9M  43.724766 -79.532242
98          M9N  43.706876 -79.518188
99          M9P  43.696319 -79.532242
100         M9R  43.688905 -79.554724
101         M9V  43.739416 -79.588437
102         M9W  43.706748 -79.594054

# Merging two tables for the langitude and longitudes of the neighbourhoods in canada data

In [266]:
df_geo=df_geo.rename({'Postal Code':'PostalCode'},axis=1)
df3 = pd.merge(df,df_geo,on='PostalCode')
df3

PostalCode                 Borough  \
0          M3A              North York   
1          M4A              North York   
2          M5A        Downtown Toronto   
3          M6A              North York   
4          M7A            Queen's Park   
5          M9A               Etobicoke   
6          M1B             Scarborough   
7          M3B              North York   
8          M4B               East York   
9          M5B        Downtown Toronto   
10         M6B              North York   
11         M9B               Etobicoke   
12         M1C             Scarborough   
13         M3C              North York   
14         M4C               East York   
15         M5C        Downtown Toronto   
16         M6C                    York   
17         M9C               Etobicoke   
18         M1E             Scarborough   
19         M4E            East Toronto   
20         M5E        Downtown Toronto   
21         M6E                    York   
22         M1G             Scarborough   
23         M4G               East York   
24         M5G        Downtown Toronto   
25         M6G        Downtown Toronto   
26         M1H             Scarborough   
27         M2H              North York   
28         M3H              North York   
29         M4H               East York   
30         M5H        Downtown Toronto   
31         M6H            West Toronto   
32         M1J             Scarborough   
33         M2J              North York   
34         M3J              North York   
35         M4J  East York/East Toronto   
36         M5J        Downtown Toronto   
37         M6J            West Toronto   
38         M1K             Scarborough   
39         M2K              North York   
40         M3K              North York   
41         M4K            East Toronto   
42         M5K        Downtown Toronto   
43         M6K            West Toronto   
44         M1L             Scarborough   
45         M2L              North York   
46         M3L              North York   
47         M4L            East Toronto   
48         M5L        Downtown Toronto   
49         M6L              North York   
50         M9L              North York   
51         M1M             Scarborough   
52         M2M              North York   
53         M3M              North York   
54         M4M            East Toronto   
55         M5M              North York   
56         M6M                    York   
57         M9M              North York   
58         M1N             Scarborough   
59         M2N              North York   
60         M3N              North York   
61         M4N         Central Toronto   
62         M5N         Central Toronto   
63         M6N                    York   
64         M9N                    York   
65         M1P             Scarborough   
66         M2P              North York   
67         M4P         Central Toronto   
68         M5P         Central Toronto   
69         M6P            West Toronto   
70         M9P               Etobicoke   
71         M1R             Scarborough   
72         M2R              North York   
73         M4R         Central Toronto   
74         M5R         Central Toronto   
75         M6R            West Toronto   
76         M7R             Mississauga   
77         M9R               Etobicoke   
78         M1S             Scarborough   
79         M4S         Central Toronto   
80         M5S        Downtown Toronto   
81         M6S            West Toronto   
82         M1T             Scarborough   
83         M4T         Central Toronto   
84         M5T        Downtown Toronto   
85         M1V             Scarborough   
86         M4V         Central Toronto   
87         M5V        Downtown Toronto   
88         M8V               Etobicoke   
89         M9V               Etobicoke   
90         M1W             Scarborough   
91         M4W        Downtown Toronto   
92         M5W  Downtown Toronto Stn A   
93         M8W               Etobicoke   
94         M9W

In [267]:
df3.shape

(103, 5)

In [268]:
address = 'Toronto, Ontario Canada'

geolocator = Nominatim(user_agent="http")
location = geolocator.geocode(address)
latitude = location.latitude
longitude = location.longitude
print('The geograpical coordinate of Toronto Canada are {}, {}.'.format(latitude, longitude))

The geograpical coordinate of Toronto Canada are 43.6534817, -79.3839347.


# Clustering and the plotting of the neighbourhoods of Canada which contain Toronto in their Borough

# Getting all the rows from the data frame which contains Toronto in their Borough.

In [269]:
df4 = df3[df3['Borough'].str.contains('Toronto',regex=False)]
df4

PostalCode                 Borough  \
2          M5A        Downtown Toronto   
9          M5B        Downtown Toronto   
15         M5C        Downtown Toronto   
19         M4E            East Toronto   
20         M5E        Downtown Toronto   
24         M5G        Downtown Toronto   
25         M6G        Downtown Toronto   
30         M5H        Downtown Toronto   
31         M6H            West Toronto   
35         M4J  East York/East Toronto   
36         M5J        Downtown Toronto   
37         M6J            West Toronto   
41         M4K            East Toronto   
42         M5K        Downtown Toronto   
43         M6K            West Toronto   
47         M4L            East Toronto   
48         M5L        Downtown Toronto   
54         M4M            East Toronto   
61         M4N         Central Toronto   
62         M5N         Central Toronto   
67         M4P         Central Toronto   
68         M5P         Central Toronto   
69         M6P            West Toronto   
73         M4R         Central Toronto   
74         M5R         Central Toronto   
75         M6R            West Toronto   
79         M4S         Central Toronto   
80         M5S        Downtown Toronto   
81         M6S            West Toronto   
83         M4T         Central Toronto   
84         M5T        Downtown Toronto   
86         M4V         Central Toronto   
87         M5V        Downtown Toronto   
91         M4W        Downtown Toronto   
92         M5W  Downtown Toronto Stn A   
96         M4X        Downtown Toronto   
97         M5X        Downtown Toronto   
99         M4Y        Downtown Toronto   
100        M7Y   East Toronto Business   

                                          Neighborhood   Latitude  Longitude  
2                            Regent Park, Harbourfront  43.654260 -79.360636  
9                             Garden District, Ryerson  43.657162 -79.378937  
15                                      St. James Town  43.651494 -79.375418  
19                                         The Beaches  43.676357 -79.293031  
20                                         Berczy Park  43.644771 -79.373306  
24                                  Central Bay Street  43.657952 -79.387383  
25                                            Christie  43.669542 -79.422564  
30                            Richmond, Adelaide, King  43.650571 -79.384568  
31                        Dufferin, Dovercourt Village  43.669005 -79.442259  
35                                  The Danforth  East  43.685347 -79.338106  
36   Harbourfront East, Union Station, Toronto Islands  43.640816 -79.381752  
37                            Little Portugal, Trinity  43.647927 -79.419750  
41                        The Danforth West, Riverdale  43.679557 -79.352188  
42            Toronto Dominion Centre, Design Exchange  43.647177 -79.381576  
43        Brockton, Parkdale Village, Exhibition Place  43.636847 -79.428191  
47                      India Bazaar, The Beaches West  43.668999 -79.315572  
48                      Commerce Court, Victoria Hotel  43.648198 -79.379817  
54                                     Studio District  43.659526 -79.340923  
61                                       Lawrence Park  43.728020 -79.388790  
62                                            Roselawn  43.711695 -79.416936  
67                                    Davisville North  43.712751 -79.390197  
68                            Forest Hill North & West  43.696948 -79.411307  
69                       High Park, The Junction South  43.661608 -79.464763  
73                                  North Toronto West  43.715383 -79.405678  
74                 The Annex, North Midtown, Yorkville  43.672710 -79.405678  
75                              Parkdale, Roncesvalles  43.648960 -79.456325  
79                                          Davisville  43.704324 -79.388790  
80                      University of Toronto, Harbord  43.662696 -79.400049  
81                              

# Visualizing all the Neighbourhoods of the above data frame using Folium

In [198]:
map_toronto = folium.Map(location=[43.651070,-79.347015],zoom_start=10)

for lat,lng,borough,neighborhood in zip(df4['Latitude'],df4['Longitude'],df4['Borough'],df4['Neighborhood']):
    label = '{}, {}'.format(neighborhood, borough)
    label = folium.Popup(label, parse_html=True)
    folium.CircleMarker(
    [lat,lng],
    radius=5,
    popup=label,
    color='blue',
    fill=True,
    fill_color='#3186cc',
    fill_opacity=0.7,
    parse_html=False).add_to(map_toronto)
map_toronto

# Using KMeans clustering for the clsutering of the neighbourhoods

In [199]:
k=5
toronto_clustering = df4.drop(['PostalCode','Borough','Neighborhood'],1)
kmeans = KMeans(n_clusters = k,random_state=0).fit(toronto_clustering)
kmeans.labels_
df4.insert(0, 'Cluster Labels', kmeans.labels_)

In [200]:
df4

Cluster Labels PostalCode                 Borough  \
2                 4        M5A        Downtown Toronto   
9                 4        M5B        Downtown Toronto   
15                4        M5C        Downtown Toronto   
19                1        M4E            East Toronto   
20                4        M5E        Downtown Toronto   
24                4        M5G        Downtown Toronto   
25                0        M6G        Downtown Toronto   
30                4        M5H        Downtown Toronto   
31                3        M6H            West Toronto   
35                1        M4J  East York/East Toronto   
36                4        M5J        Downtown Toronto   
37                0        M6J            West Toronto   
41                1        M4K            East Toronto   
42                4        M5K        Downtown Toronto   
43                0        M6K            West Toronto   
47                1        M4L            East Toronto   
48                4        M5L        Downtown Toronto   
54                1        M4M            East Toronto   
61                2        M4N         Central Toronto   
62                2        M5N         Central Toronto   
67                2        M4P         Central Toronto   
68                2        M5P         Central Toronto   
69                3        M6P            West Toronto   
73                2        M4R         Central Toronto   
74                0        M5R         Central Toronto   
75                3        M6R            West Toronto   
79                2        M4S         Central Toronto   
80                0        M5S        Downtown Toronto   
81                3        M6S            West Toronto   
83                2        M4T         Central Toronto   
84                0        M5T        Downtown Toronto   
86                2        M4V         Central Toronto   
87                0        M5V        Downtown Toronto   
91                4        M4W        Downtown Toronto   
92                4        M5W  Downtown Toronto Stn A   
96                4        M4X        Downtown Toronto   
97                4        M5X        Downtown Toronto   
99                4        M4Y        Downtown Toronto   
100               1        M7Y   East Toronto Business   

                                          Neighborhood   Latitude  Longitude  
2                            Regent Park, Harbourfront  43.654260 -79.360636  
9                             Garden District, Ryerson  43.657162 -79.378937  
15                                      St. James Town  43.651494 -79.375418  
19                                         The Beaches  43.676357 -79.293031  
20                                         Berczy Park  43.644771 -79.373306  
24                                  Central Bay Street  43.657952 -79.387383  
25                                            Christie  43.669542 -79.422564  
30                            Richmond, Adelaide, King  43.650571 -79.384568  
31                        Dufferin, Dovercourt Village  43.669005 -79.442259  
35                                  The Danforth  East  43.685347 -79.338106  
36   Harbourfront East, Union Station, Toronto Islands  43.640816 -79.381752  
37                            Little Portugal, Trinity  43.647927 -79.419750  
41                        The Danforth West, Riverdale  43.679557 -79.352188  
42            Toronto Dominion Centre, Design Exchange  43.647177 -79.381576  
43        Brockton, Parkdale Village, Exhibition Place  43.636847 -79.428191  
47                      India Bazaar, The Beaches West  43.668999 -79.315572  
48                      Commerce Court, Victoria Hotel  43.648198 -79.379817  
54                                     Studio District  43.659526 -79.340923  
61                                       Lawrence Park  43.728020 -79.388790  
62                                            Roselawn  43.711695 -79.416936  
67                       

In [201]:
# create map
map_clusters = folium.Map(location=[43.651070,-79.347015],zoom_start=10)

# set color scheme for the clusters
x = np.arange(k)
ys = [i + x + (i*x)**2 for i in range(k)]
colors_array = cm.rainbow(np.linspace(0, 1, len(ys)))
rainbow = [colors.rgb2hex(i) for i in colors_array]

# add markers to the map
markers_colors = []
for lat, lon, neighbourhood, cluster in zip(df4['Latitude'], df4['Longitude'], df4['Neighborhood'], df4['Cluster Labels']):
    label = folium.Popup(' Cluster ' + str(cluster), parse_html=True)
    folium.CircleMarker(
        [lat, lon],
        radius=5,
        popup=label,
        color=rainbow[cluster-1],
        fill=True,
        fill_color=rainbow[cluster-1],
        fill_opacity=0.7).add_to(map_clusters)
       
map_clusters

# Utilizing foursquare API to explore and segment neighborhoods

In [202]:
CLIENT_ID = 'HWUAEI4NZJ5SNUPRXKO2AGDKKPJ2AVOTHLGGDJTAIF3WDYUW' # your Foursquare ID
CLIENT_SECRET = '2M5J4LSZABRGEQUIOS1EU3RSO0JOPVZ3TAWZM0AYPZUJMYP5' # your Foursquare Secret
ACCESS_TOKEN = 'JL3KMOV1KGSUYRLR1QYCHJVPJHSFC3MMN41JKBDG5DAX2UY1' # your FourSquare Access Token
VERSION = '20180604'
LIMIT = 30
print('Your credentails:')
print('CLIENT_ID: ' + CLIENT_ID)
print('CLIENT_SECRET:' + CLIENT_SECRET)

Your credentails:
CLIENT_ID: HWUAEI4NZJ5SNUPRXKO2AGDKKPJ2AVOTHLGGDJTAIF3WDYUW
CLIENT_SECRET:2M5J4LSZABRGEQUIOS1EU3RSO0JOPVZ3TAWZM0AYPZUJMYP5


# Exploring neighborhood in Toronto

In [350]:
def getNearbyVenues(names, latitudes, longitudes, radius=500):
    
    venues_list=[]
    for name, lat, lng in zip(names, latitudes, longitudes):
        print(name)
        
        url = 'https://api.foursquare.com/v2/venues/explore?&client_id=HWUAEI4NZJ5SNUPRXKO2AGDKKPJ2AVOTHLGGDJTAIF3WDYUW&client_secret=2M5J4LSZABRGEQUIOS1EU3RSO0JOPVZ3TAWZM0AYPZUJMYP5&v=20180604&ll=43.668999,-79.315572&radius=500&limit=30'.format(
            CLIENT_ID, CLIENT_SECRET, VERSION, lat, lng, radius, LIMIT)
        
        results = requests.get(url).json()["response"]['groups'][0]['items']
        
        venues_list.append([( name, lat, lng, v['venue']['name'], v['venue']['location']['lat'], v['venue']['location']['lng'], v['venue']['categories'][0]['name']) for v in results])

    nearby_venues = pd.DataFrame([item for venue_list in venues_list for item in venue_list])
    nearby_venues.columns = ['Neighborhood', 'Neighborhood Latitude', 'Neighborhood Longitude', 'Venue', 'Venue Latitude', 'Venue Longitude', 'Venue Category']
    
    return(nearby_venues)

# Retreiving all Venues

In [351]:
toronto_neighborhoods = df4
toronto_venues = getNearbyVenues(names=toronto_neighborhoods['Neighborhood'],
                                   latitudes=toronto_neighborhoods['Latitude'],
                                   longitudes=toronto_neighborhoods['Longitude']
                                  )

Regent Park, Harbourfront
Garden District, Ryerson
St. James Town
The Beaches
Berczy Park
Central Bay Street
Christie
Richmond, Adelaide, King
Dufferin, Dovercourt Village
The Danforth  East
Harbourfront East, Union Station, Toronto Islands
Little Portugal, Trinity
The Danforth West, Riverdale
Toronto Dominion Centre, Design Exchange
Brockton, Parkdale Village, Exhibition Place
India Bazaar, The Beaches West
Commerce Court, Victoria Hotel
Studio District
Lawrence Park
Roselawn
Davisville North
Forest Hill North & West
High Park, The Junction South
North Toronto West
The Annex, North Midtown, Yorkville
Parkdale, Roncesvalles
Davisville
University of Toronto, Harbord
Runnymede, Swansea
Moore Park, Summerhill East
Kensington Market, Chinatown, Grange Park
Summerhill West, Rathnelly, South Hill, Forest Hill SE, Deer Park
CN Tower, King and Spadina, Railway Lands, Harbourfront West, Bathurst Quay, South Niagara, Island airport
Rosedale
Enclave of M5E
St. James Town, Cabbagetown
First Canadi

In [352]:
print(toronto_venues.shape)
toronto_venues.head()

(702, 7)


Neighborhood  Neighborhood Latitude  Neighborhood Longitude  \
0  Regent Park, Harbourfront               43.65426              -79.360636   
1  Regent Park, Harbourfront               43.65426              -79.360636   
2  Regent Park, Harbourfront               43.65426              -79.360636   
3  Regent Park, Harbourfront               43.65426              -79.360636   
4  Regent Park, Harbourfront               43.65426              -79.360636   

                        Venue  Venue Latitude  Venue Longitude  \
0              System Fitness       43.667171       -79.312733   
1  British Style Fish & Chips       43.668723       -79.317139   
2           Brett's Ice Cream       43.667222       -79.312831   
3         The Burger's Priest       43.666731       -79.315556   
4                     O Sushi       43.666684       -79.316614   

         Venue Category  
0                   Gym  
1     Fish & Chips Shop  
2        Ice Cream Shop  
3  Fast Food Restaurant  
4      Sushi Restaurant

In [353]:
toronto_venues.groupby('Neighborhood').count()

Neighborhood Latitude  \
Neighborhood                                                                
Berczy Park                                                            18   
Brockton, Parkdale Village, Exhibition Place                           18   
CN Tower, King and Spadina, Railway Lands, Harb...                     18   
Central Bay Street                                                     18   
Christie                                                               18   
Church and Wellesley                                                   18   
Commerce Court, Victoria Hotel                                         18   
Davisville                                                             18   
Davisville North                                                       18   
Dufferin, Dovercourt Village                                           18   
Enclave of M4L                                                         18   
Enclave of M5E                                                         18   
First Canadian Place, Underground city                                 18   
Forest Hill North & West                                               18   
Garden District, Ryerson                                               18   
Harbourfront East, Union Station, Toronto Islands                      18   
High Park, The Junction South                                          18   
India Bazaar, The Beaches West                                         18   
Kensington Market, Chinatown, Grange Park                              18   
Lawrence Park                                                          18   
Little Portugal, Trinity                                               18   
Moore Park, Summerhill East                                            18   
North Toronto West                                                     18   
Parkdale, Roncesvalles                                                 18   
Regent Park, Harbourfront                                              18   
Richmond, Adelaide, King                                               18   
Rosedale                                                               18   
Roselawn                                                               18   
Runnymede, Swansea                                                     18   
St. James Town                                                         18   
St. James Town, Cabbagetown                                            18   
Studio District                                                        18   
Summerhill West, Rathnelly, South Hill, Forest ...                     18   
The Annex, North Midtown, Yorkville                                    18   
The Beaches                                                            18   
The Danforth  East                                                     18   
The Danforth West, Riverdale                                           18   
Toronto Dominion Centre, Design Exchange                               18   
University of Toronto, Harbord                                         18   

                                                    Neighborhood Longitude  \
Neighborhood                                                                 
Berczy Park                                                             18   
Brockton, Parkdale Village, Exhibition Place                            18   
CN Tower, King and Spadina, Railway Lands, Harb...                      18   
Central Bay Street                                                      18   
Christie                                                                18   
Church and Wellesley                                                    18   
Commerce Court, Victoria Hotel                                          18   
Davisville                                                              18   
Davisville North                                                        18   
Dufferin, Dovercourt Village        

In [354]:
print('There are {} uniques categories.'.format(len(toronto_venues['Venue Category'].unique())))

There are 18 uniques categories.


# Analyzing each borough neighborhood

In [355]:
toronto_onehot = pd.get_dummies(toronto_venues[['Venue Category']], prefix="", prefix_sep="")

toronto_onehot['Neighborhood'] = toronto_venues['Neighborhood'] 

fixed_columns = [toronto_onehot.columns[-1]] + list(toronto_onehot.columns[:-1])
toronto_onehot = toronto_onehot[fixed_columns]

toronto_onehot.head()

Neighborhood  Board Shop  Brewery  Burrito Place  \
0  Regent Park, Harbourfront           0        0              0   
1  Regent Park, Harbourfront           0        0              0   
2  Regent Park, Harbourfront           0        0              0   
3  Regent Park, Harbourfront           0        0              0   
4  Regent Park, Harbourfront           0        0              0   

   Fast Food Restaurant  Fish & Chips Shop  Gym  Ice Cream Shop  \
0                     0                  0    1               0   
1                     0                  1    0               0   
2                     0                  0    0               1   
3                     1                  0    0               0   
4                     0                  0    0               0   

   Italian Restaurant  Liquor Store  Movie Theater  Park  Pet Store  \
0                   0             0              0     0          0   
1                   0             0              0     0          0   
2                   0             0              0     0          0   
3                   0             0              0     0          0   
4                   0             0              0     0          0   

   Pizza Place  Pub  Restaurant  Sandwich Place  Steakhouse  Sushi Restaurant  
0            0    0           0               0           0                 0  
1            0    0           0               0           0                 0  
2            0    0           0               0           0                 0  
3            0    0           0               0           0                 0  
4            0    0           0               0           0                 1

In [356]:
toronto_onehot.shape

(702, 19)

In [357]:
toronto_grouped = toronto_onehot.groupby('Neighborhood').mean().reset_index()
toronto_grouped

Neighborhood  Board Shop   Brewery  \
0                                         Berczy Park    0.055556  0.055556   
1        Brockton, Parkdale Village, Exhibition Place    0.055556  0.055556   
2   CN Tower, King and Spadina, Railway Lands, Har...    0.055556  0.055556   
3                                  Central Bay Street    0.055556  0.055556   
4                                            Christie    0.055556  0.055556   
5                                Church and Wellesley    0.055556  0.055556   
6                      Commerce Court, Victoria Hotel    0.055556  0.055556   
7                                          Davisville    0.055556  0.055556   
8                                    Davisville North    0.055556  0.055556   
9                        Dufferin, Dovercourt Village    0.055556  0.055556   
10                                     Enclave of M4L    0.055556  0.055556   
11                                     Enclave of M5E    0.055556  0.055556   
12             First Canadian Place, Underground city    0.055556  0.055556   
13                           Forest Hill North & West    0.055556  0.055556   
14                           Garden District, Ryerson    0.055556  0.055556   
15  Harbourfront East, Union Station, Toronto Islands    0.055556  0.055556   
16                      High Park, The Junction South    0.055556  0.055556   
17                     India Bazaar, The Beaches West    0.055556  0.055556   
18          Kensington Market, Chinatown, Grange Park    0.055556  0.055556   
19                                      Lawrence Park    0.055556  0.055556   
20                           Little Portugal, Trinity    0.055556  0.055556   
21                        Moore Park, Summerhill East    0.055556  0.055556   
22                                 North Toronto West    0.055556  0.055556   
23                             Parkdale, Roncesvalles    0.055556  0.055556   
24                          Regent Park, Harbourfront    0.055556  0.055556   
25                           Richmond, Adelaide, King    0.055556  0.055556   
26                                           Rosedale    0.055556  0.055556   
27                                           Roselawn    0.055556  0.055556   
28                                 Runnymede, Swansea    0.055556  0.055556   
29                                     St. James Town    0.055556  0.055556   
30                        St. James Town, Cabbagetown    0.055556  0.055556   
31                                    Studio District    0.055556  0.055556   
32  Summerhill West, Rathnelly, South Hill, Forest...    0.055556  0.055556   
33                The Annex, North Midtown, Yorkville    0.055556  0.055556   
34                                        The Beaches    0.055556  0.055556   
35                                 The Danforth  East    0.055556  0.055556   
36                       The Danforth West, Riverdale    0.055556  0.055556   
37           Toronto Dominion Centre, Design Exchange    0.055556  0.055556   
38                     University of Toronto, Harbord    0.055556  0.055556   

    Burrito Place  Fast Food Restaurant  Fish & Chips Shop       Gym  \
0        0.055556              0.055556           0.055556  0.055556   
1        0.055556              0.055556           0.055556  0.055556   
2        0.055556              0.055556           0.055556  0.055556   
3        0.055556              0.055556           0.055556  0.055556   
4        0.055556              0.055556           0.055556  0.055556   
5        0.055556              0.055556           0.055556  0.055556   
6        0.055556              0.055556           0.055556  0.055556   
7        0.055556              0.055556           0.055556  0.055556   
8        0.055556              0.055556           0.055556  0.055556   
9        0.055556              0.055556           0.055556  0.055556   
10       0.055556              0.055556           0.055556  0.055556   
11       0.05555

In [358]:
toronto_grouped.shape

(39, 19)

In [359]:
num_top_venues = 5
for neigh in toronto_grouped['Neighborhood']:
    print("----"+neigh+"----")
    temp = toronto_grouped[toronto_grouped['Neighborhood'] == neigh].T.reset_index()
    temp.columns = ['venue','freq']
    temp = temp.iloc[1:]
    temp['freq'] = temp['freq'].astype(float)
    temp = temp.round({'freq': 2})
    print(temp.sort_values('freq', ascending=False).reset_index(drop=True).head(num_top_venues))
    print('\n')

----Berczy Park----
            venue  freq
0      Board Shop  0.06
1         Brewery  0.06
2      Steakhouse  0.06
3  Sandwich Place  0.06
4      Restaurant  0.06


----Brockton, Parkdale Village, Exhibition Place----
            venue  freq
0      Board Shop  0.06
1         Brewery  0.06
2      Steakhouse  0.06
3  Sandwich Place  0.06
4      Restaurant  0.06


----CN Tower, King and Spadina, Railway Lands, Harbourfront West, Bathurst Quay, South Niagara, Island airport----
            venue  freq
0      Board Shop  0.06
1         Brewery  0.06
2      Steakhouse  0.06
3  Sandwich Place  0.06
4      Restaurant  0.06


----Central Bay Street----
            venue  freq
0      Board Shop  0.06
1         Brewery  0.06
2      Steakhouse  0.06
3  Sandwich Place  0.06
4      Restaurant  0.06


----Christie----
            venue  freq
0      Board Shop  0.06
1         Brewery  0.06
2      Steakhouse  0.06
3  Sandwich Place  0.06
4      Restaurant  0.06


----Church and Wellesley----
         

In [360]:
def return_most_common_venues(row, num_top_venues):
    row_categories = row.iloc[1:]
    row_categories_sorted = row_categories.sort_values(ascending=False)
    return row_categories_sorted.index.values[0:num_top_venues]

In [361]:
import numpy as np
num_top_venues = 10
indicators = ['st', 'nd', 'rd']

columns = ['Neighborhood']
for ind in np.arange(num_top_venues):
    try:
        columns.append('{}{} Most Common Venue'.format(ind+1, indicators[ind]))
    except:
        columns.append('{}th Most Common Venue'.format(ind+1))

neighborhoods_venues_sorted = pd.DataFrame(columns=columns)
neighborhoods_venues_sorted['Neighborhood'] = toronto_grouped['Neighborhood']

for ind in np.arange(toronto_grouped.shape[0]):
    neighborhoods_venues_sorted.iloc[ind, 1:] = return_most_common_venues(toronto_grouped.iloc[ind, :], num_top_venues)

neighborhoods_venues_sorted.shape

(39, 11)

# Clustering Neighborhood

In [362]:
from sklearn.cluster import KMeans

km = KMeans(n_clusters=3, init='k-means++', max_iter=100, n_init=1, 
  verbose=True)

In [373]:
# set number of clusters
kclusters = 5

toronto_grouped_clustering = toronto_grouped.drop("Neighborhood", 1)

# run k-means clustering
kmeans = KMeans(n_clusters=kclusters, random_state=0).fit(toronto_grouped_clustering)

# check cluster labels generated for each row in the dataframe
kmeans.labels_[0:10]

/Library/Frameworks/Python.framework/Versions/3.7/lib/python3.7/site-packages/ipykernel_launcher.py:7: ConvergenceWarning: Number of distinct clusters (1) found smaller than n_clusters (5). Possibly due to duplicate points in X.
  import sys


array([0, 0, 0, 0, 0, 0, 0, 0, 0, 0], dtype=int32)

In [374]:
toronto_neighborhoods

PostalCode                 Borough  \
2          M5A        Downtown Toronto   
9          M5B        Downtown Toronto   
15         M5C        Downtown Toronto   
19         M4E            East Toronto   
20         M5E        Downtown Toronto   
24         M5G        Downtown Toronto   
25         M6G        Downtown Toronto   
30         M5H        Downtown Toronto   
31         M6H            West Toronto   
35         M4J  East York/East Toronto   
36         M5J        Downtown Toronto   
37         M6J            West Toronto   
41         M4K            East Toronto   
42         M5K        Downtown Toronto   
43         M6K            West Toronto   
47         M4L            East Toronto   
48         M5L        Downtown Toronto   
54         M4M            East Toronto   
61         M4N         Central Toronto   
62         M5N         Central Toronto   
67         M4P         Central Toronto   
68         M5P         Central Toronto   
69         M6P            West Toronto   
73         M4R         Central Toronto   
74         M5R         Central Toronto   
75         M6R            West Toronto   
79         M4S         Central Toronto   
80         M5S        Downtown Toronto   
81         M6S            West Toronto   
83         M4T         Central Toronto   
84         M5T        Downtown Toronto   
86         M4V         Central Toronto   
87         M5V        Downtown Toronto   
91         M4W        Downtown Toronto   
92         M5W  Downtown Toronto Stn A   
96         M4X        Downtown Toronto   
97         M5X        Downtown Toronto   
99         M4Y        Downtown Toronto   
100        M7Y   East Toronto Business   

                                          Neighborhood   Latitude  Longitude  \
2                            Regent Park, Harbourfront  43.654260 -79.360636   
9                             Garden District, Ryerson  43.657162 -79.378937   
15                                      St. James Town  43.651494 -79.375418   
19                                         The Beaches  43.676357 -79.293031   
20                                         Berczy Park  43.644771 -79.373306   
24                                  Central Bay Street  43.657952 -79.387383   
25                                            Christie  43.669542 -79.422564   
30                            Richmond, Adelaide, King  43.650571 -79.384568   
31                        Dufferin, Dovercourt Village  43.669005 -79.442259   
35                                  The Danforth  East  43.685347 -79.338106   
36   Harbourfront East, Union Station, Toronto Islands  43.640816 -79.381752   
37                            Little Portugal, Trinity  43.647927 -79.419750   
41                        The Danforth West, Riverdale  43.679557 -79.352188   
42            Toronto Dominion Centre, Design Exchange  43.647177 -79.381576   
43        Brockton, Parkdale Village, Exhibition Place  43.636847 -79.428191   
47                      India Bazaar, The Beaches West  43.668999 -79.315572   
48                      Commerce Court, Victoria Hotel  43.648198 -79.379817   
54                                     Studio District  43.659526 -79.340923   
61                                       Lawrence Park  43.728020 -79.388790   
62                                            Roselawn  43.711695 -79.416936   
67                                    Davisville North  43.712751 -79.390197   
68                            Forest Hill North & West  43.696948 -79.411307   
69                       High Park, The Junction South  43.661608 -79.464763   
73                                  North Toronto West  43.715383 -79.405678   
74                 The Annex, North Midtown, Yorkville  43.672710 -79.405678   
75                              Parkdale, Roncesvalles  43.648960 -79.456325   
79                                          Davisville  43.704324 -79.388790   
80                      University of Toronto, Harbord  43.662696 -79.400049   
81 

In [375]:
toronto_merged = toronto_neighborhoods

# add clustering labels
toronto_merged['Cluster Labels'] = kmeans.labels_

# merge toronto_grouped with toronto_data to add latitude/longitude for each neighborhood
toronto_merged = toronto_merged.join(neighborhoods_venues_sorted.set_index('Neighborhood'), on='Neighborhood')
toronto_merged

/Library/Frameworks/Python.framework/Versions/3.7/lib/python3.7/site-packages/ipykernel_launcher.py:4: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: http://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  after removing the cwd from sys.path.


PostalCode                 Borough  \
2          M5A        Downtown Toronto   
9          M5B        Downtown Toronto   
15         M5C        Downtown Toronto   
19         M4E            East Toronto   
20         M5E        Downtown Toronto   
24         M5G        Downtown Toronto   
25         M6G        Downtown Toronto   
30         M5H        Downtown Toronto   
31         M6H            West Toronto   
35         M4J  East York/East Toronto   
36         M5J        Downtown Toronto   
37         M6J            West Toronto   
41         M4K            East Toronto   
42         M5K        Downtown Toronto   
43         M6K            West Toronto   
47         M4L            East Toronto   
48         M5L        Downtown Toronto   
54         M4M            East Toronto   
61         M4N         Central Toronto   
62         M5N         Central Toronto   
67         M4P         Central Toronto   
68         M5P         Central Toronto   
69         M6P            West Toronto   
73         M4R         Central Toronto   
74         M5R         Central Toronto   
75         M6R            West Toronto   
79         M4S         Central Toronto   
80         M5S        Downtown Toronto   
81         M6S            West Toronto   
83         M4T         Central Toronto   
84         M5T        Downtown Toronto   
86         M4V         Central Toronto   
87         M5V        Downtown Toronto   
91         M4W        Downtown Toronto   
92         M5W  Downtown Toronto Stn A   
96         M4X        Downtown Toronto   
97         M5X        Downtown Toronto   
99         M4Y        Downtown Toronto   
100        M7Y   East Toronto Business   

                                          Neighborhood   Latitude  Longitude  \
2                            Regent Park, Harbourfront  43.654260 -79.360636   
9                             Garden District, Ryerson  43.657162 -79.378937   
15                                      St. James Town  43.651494 -79.375418   
19                                         The Beaches  43.676357 -79.293031   
20                                         Berczy Park  43.644771 -79.373306   
24                                  Central Bay Street  43.657952 -79.387383   
25                                            Christie  43.669542 -79.422564   
30                            Richmond, Adelaide, King  43.650571 -79.384568   
31                        Dufferin, Dovercourt Village  43.669005 -79.442259   
35                                  The Danforth  East  43.685347 -79.338106   
36   Harbourfront East, Union Station, Toronto Islands  43.640816 -79.381752   
37                            Little Portugal, Trinity  43.647927 -79.419750   
41                        The Danforth West, Riverdale  43.679557 -79.352188   
42            Toronto Dominion Centre, Design Exchange  43.647177 -79.381576   
43        Brockton, Parkdale Village, Exhibition Place  43.636847 -79.428191   
47                      India Bazaar, The Beaches West  43.668999 -79.315572   
48                      Commerce Court, Victoria Hotel  43.648198 -79.379817   
54                                     Studio District  43.659526 -79.340923   
61                                       Lawrence Park  43.728020 -79.388790   
62                                            Roselawn  43.711695 -79.416936   
67                                    Davisville North  43.712751 -79.390197   
68                            Forest Hill North & West  43.696948 -79.411307   
69                       High Park, The Junction South  43.661608 -79.464763   
73                                  North Toronto West  43.715383 -79.405678   
74                 The Annex, North Midtown, Yorkville  43.672710 -79.405678   
75                              Parkdale, Roncesvalles  43.648960 -79.456325   
79                                          Davisville  43.704324 -79.388790   
80                      University of Toronto, Harbord  43.662696 -79.400049   
81 

# Visualizing Neighborhood

In [376]:

# create map
map_clusters = folium.Map(location=[Latitude, Longitude], zoom_start=11)

# set color scheme for the clusters
x = np.arange(kclusters)
ys = [i+x+(i*x)**2 for i in range(kclusters)]
colors_array = cm.rainbow(np.linspace(0, 1, len(ys)))
rainbow = [colors.rgb2hex(i) for i in colors_array]

# add markers to the map
markers_colors = []
for Latitude, Longitude, poi, cluster in zip(toronto_merged['Latitude'], toronto_merged['Longitude'], toronto_merged['Neighborhood'],kmeans.labels_):
    label = folium.Popup(str(poi) + ' Cluster ' + str(cluster), parse_html=True)
    folium.CircleMarker(
        [latitude, longitude],
        radius=5,
        popup=label,
        color=rainbow[cluster-1],
        fill=True,
        fill_color=rainbow[cluster-1],
        fill_opacity=0.7).add_to(map_clusters)
       
map_clusters

# Conclusion
After analyzing all the neighbourhoods in the borough M4L borough came with most number of recreational places nearby and hence its a best place to live in toronto.

# Applied Capstone Project
## Table of contents
* [Introduction: Business Problem](#1)
* [Data](#2)
* [Methodology](#3)
* [Analysis](#4)
* [Results and Discussion](#5)
* [Conclusion](#6)

# 1. Intorduction <a name="1"></a> 
The purpose of this Capstone Project is to help people in exploring better facilities around their neighborhood. It will help people making smart and efficient decision on selecting great neighborhood out of numbers of other neighborhoods in Scarborough, Toronto. This project is for those people who are looking for better neighborhoods. For ease of accessing to Cafe, School, Super market, medical shops, grocery shops, mall, theater, hospital, like minded people, etc. This Capstone Project aim to create an analysis of features for a people migrating to Scarborough to search a best neighborhood as a comparative analysis between neighborhoods. It will help people to get awareness of the area and neighborhood before moving to a new city, state, country or place for their work or to start a new fresh life.

# 2. Data <a name="2"></a> 
Data Link: https://en.wikipedia.org/wiki/List_of_postal_codes_of_Canada:_M Will use Scarborough dataset which we scrapped from Wikipedia on Week 3. Dataset consisting of latitude and longitude, zip codes.

## Foursquare API Data:
We will need data about different venues in different neighborhoods of that specific borough. In order to gain that information we will use “Foursquare” location information. Foursquare is a location data provider with information about all manner of venues and events within an area of interest. Such information includes venue names, locations, menus and even photos. As such, the foursquare location platform will be used as the sole data source since all the stated required information can be obtained through the API. After finding the list of neighborhoods, we then connect to the Foursquare API to gather information about venues inside each and every neighborhood. For each neighborhood, we have chosen the radius to be 100 meter. The data retrieved from Foursquare contained information of venues within a specified distance of the longitude and latitude of the postcodes. The information obtained per venue as follows:

* Neighborhood
* Neighborhood Latitude
* Neighborhood Longitude
* Venue
* Name of the venue e.g. the name of a store or restaurant
* Venue Latitude
* Venue Longitude
* Venue Category

# 3. Methodology Section <a name="3"></a>
## Clustering Approach:
To compare the similarities of two cities, we decided to explore neighborhoods, segment them, and group them into clusters to find similar neighborhoods in a big city like New York and Toronto. To be able to do that, we need to cluster data which is a form of unsupervised machine learning: k-means clustering algorithm. Using K-Means Clustering Approach | Most Common Venue Most Common Venues near Neighborhood | Using Clustering

## Work Flow:
Using credentials of Foursquare API features of near-by places of the neighborhoods would be mined. Due to http request limitations the number of places per neighborhood parameter would reasonably be set to 100 and the radius parameter would be set to 500.

# 4. Analysis Section <a name="4"></a> 
Map of Clusters in Scarborough Average Housing Price by Clusters in Scarborough School Ratings by Clusters in Scarborough

## The Location:
Scarborough is a popular destination for new immigrants in Canada to reside. As a result, it is one of the most diverse and multicultural areas in the Greater Toronto Area, being home to various religious groups and places of worship. Although immigration has become a hot topic over the past few years with more governments seeking more restrictions on immigrants and refugees, the general trend of immigration into Canada has been one of on the rise.

## Foursquare API:
This Capstone project have used Four-square API as its prime data gathering source as it has a database of millions of places, especially their places API which provides the ability to perform location search, location sharing and details about a business.

# 5. Discussion Section <a name="5"></a> 
## Problem Which Tried to Solve:
The major purpose of this project, is to suggest a better neighborhood in a new city for the person who are shifting there. Social presence in society in terms of like minded people. Connectivity to the airport, bus stand, city center, markets and other daily needs things nearby.

# 6. Conclusion Section <a name="6"></a> 
In this Capstone project, using k-means cluster algorithm I separated the neighborhood into 5 different clusters and for 103 different latitude and longitude from dataset, which have very-similar neighborhoods around them. On the basis of results presented in this notebook a particular neighborhood based on the venues nearby would be suitable for a new person moving to Toronto will be.

## Libraries Which are Used to Develop the Project:
Pandas pandas: For creating and manipulating dataframes
Folium: Python visualization library would be used to visualize the neighborhoods cluster distribution of using interactive leaflet map.
Scikit Learn: For importing k-means clustering.
XML: To separate data from presentation and XML stores data in plain text format.
Geocoder: To retrieve Location Data.
Beautiful Soup and Requests: To scrap and library to handle http requests.
​
